crates/composition/src/emit_federated_graph.rs

```diff
@@ -129,3 +144,70 @@ fn emit_union_members(
         }
     }
 }
+
+fn emit_keys(keys: &[KeyIr], ctx: &mut Context<'_>) {
+    for KeyIr { object_id, key_id } in keys {
+        let parent_id = federated::Definition::Object(*object_id);
+        let key = ctx.subgraphs.walk(*key_id);
+        let selection = attach_selection(key.fields(), parent_id, ctx);
+        ctx.out[*object_id].resolvable_keys.push(federated::Key {
+            subgraph_id: federated::SubgraphId(key.parent_definition().subgraph().id.idx()),
+            fields: selection,
+        });
+    }
+}
+
+/// Attach a selection set defined in strings to a FederatedGraph, transforming the strings into
+/// field ids.
+fn attach_selection(
+    selection_set: &[subgraphs::Selection],
+    parent_id: federated::Definition,
+    ctx: &mut Context<'_>,
+) -> federated::SelectionSet {
+    selection_set
+        .iter()
+        .map(|selection| {
+            let field = match parent_id {
+                federated::Definition::Object(object_id) => ctx
+                    .out
+                    .object_fields
+                    .iter()
+                    .find(|object_field| {
+                        let field_name = ctx.out[object_field.field_id].name;
+                        object_field.object_id == object_id
+                            && ctx.strings_ir[field_name]
+                                == ctx.subgraphs.walk(selection.field).as_str()
+                    })
+                    .map(|of| of.field_id)
+                    .unwrap(),
```

Comment:
Ugly and inefficient. But we're trying to iterate fast, so I decided not to improve. If you feel it would be better, I'm glad to do it.

Reply:
Yep, that's the case since `emit_keys` comes after `emit_fields` (and the objects are already inserted earlier during composition). The Vecs in FederatedGraph aren't sorted, the only guarantee is that the ids are correct. Everything is produced already valid, there's no binary search involved.

Reply:
if ids are correct, should the `object_id` be enough to find the right fields?

Reply:
So in the scope of this function we have:

1. The federated graph with all fields and objects
2. The object (or interface) id to which the selection set applies to
3. A `StringId` for the selection (the name of the selected field)

What we do here is find find an `ObjectField` (or `InterfaceField`) in (1.) where (2.) is the ObjectField::object_id and (3.) is the name of field referenced by ObjectField::field_id.

Where we could be more efficient is in finding the right `ObjectField`s. Maybe a binary search on the ObjectField::object_id could work to narrow down, but I'm not sure we can guarantee the that the vec is sorted here — I can check. The other solution would be building some kind of index (`HashMap<(ObjectId, StringId), FieldId>` or similar).

Reply:
mixed the structure, thought for a moment that it was `ObjectFields { object_id, fields: Vec<FieldId> }`. I would sort the vec to generate the keys. 🤷 
It doesn't need to be part of the contract.

Reply:
Fixed in 058cff7d5048b9d3a989610349227a538411bf28